[ValueTracking] Fix "getOperand() out of range!" assertion crash

vedantparanjape-amd opened the pull request:
Return instruction can also return void in LLVM IR. In such cases the number of operands are zero. So, call to getOperand(0) will result in an assertion failure as well.

Instead of using getOperand(0), we use getReturnValue() as it graciously return nullptr and prevents a crash. This bug was triggered in inliner, and this patch also adds a testcase. This patch fixes the issue #87441.

vedantparanjape-amd committed:
[ValueTracking] Fix "getOperand() out of range!" assertion crash

Return instruction can also return void in LLVM IR. In such cases the
number of operands are zero. So, call to getOperand(0) will result in an
assertion failure as well.

Instead of using getOperand(0), we use getReturnValue() as it graciously
return nullptr and prevents a crash. This bug was triggered in inliner,
and this patch also adds a testcase. This patch fixes the issue #87441.

preames commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -7378,7 +7378,7 @@ static bool handleGuaranteedWellDefinedOps(const Instruction *I,
     case Instruction::Ret:
       if (I->getFunction()->hasRetAttribute(Attribute::NoUndef) &&
-          Handle(I->getOperand(0)))
+          Handle(cast<ReturnInst>(I)->getReturnValue()))
```
Calling Handle with a nullptr seems surprising here.  I'd add a clause to the if condition to avoid calling Handle instead.

vedantparanjape-amd commented:
> Not sure this is really the right fix. If a function has a noundef attribute on the return value, then it cannot have a void return. This looks more like a context problem, in that we're dealing with two different functions during inlining.

I suspected this, but the LangRef for noundef isn't really clear on this aspect: https://llvm.org/docs/LangRef.html. It's not mentioned anywhere that noundef attribute on the return value means it can't have a void return.

> 
> Maybe it's an acceptable workaround, but I'm not convinced it really addresses all issues that can arise from this.

nikic commented:
> > Not sure this is really the right fix. If a function has a noundef attribute on the return value, then it cannot have a void return. This looks more like a context problem, in that we're dealing with two different functions during inlining.
> 
> I suspected this, but the LangRef for noundef isn't really clear on this aspect: https://llvm.org/docs/LangRef.html. It's not mentioned anywhere that noundef attribute on the return value means it can't have a void return.

Yeah, this is not spelled out in LangRef, but it's verified here: https://github.com/llvm/llvm-project/blob/4ae33c52f794dbd64924dd006570cdc409c297bc/llvm/lib/IR/Attributes.cpp#L2067-L2071

vedantparanjape-amd commented:
> > > Not sure this is really the right fix. If a function has a noundef attribute on the return value, then it cannot have a void return. This looks more like a context problem, in that we're dealing with two different functions during inlining.
> > 
> > 
> > I suspected this, but the LangRef for noundef isn't really clear on this aspect: https://llvm.org/docs/LangRef.html. It's not mentioned anywhere that noundef attribute on the return value means it can't have a void return.
> 
> Yeah, this is not spelled out in LangRef, but it's verified here:
> 
> https://github.com/llvm/llvm-project/blob/4ae33c52f794dbd64924dd006570cdc409c297bc/llvm/lib/IR/Attributes.cpp#L2067-L2071

I see, this fix is just a workaround. noundef should not apply to the inlined function as such, as in atleast for ret instructions.

vedantparanjape-amd commented:
> > > > Not sure this is really the right fix. If a function has a noundef attribute on the return value, then it cannot have a void return. This looks more like a context problem, in that we're dealing with two different functions during inlining.
> > > 
> > > 
> > > I suspected this, but the LangRef for noundef isn't really clear on this aspect: https://llvm.org/docs/LangRef.html. It's not mentioned anywhere that noundef attribute on the return value means it can't have a void return.
> > 
> > 
> > Yeah, this is not spelled out in LangRef, but it's verified here:
> > https://github.com/llvm/llvm-project/blob/4ae33c52f794dbd64924dd006570cdc409c297bc/llvm/lib/IR/Attributes.cpp#L2067-L2071
> 
> I see, this fix is just a workaround. noundef should not apply to the inlined function as such, as in atleast for ret instructions.

vedantparanjape-amd commented:
@nikic ping!

nikic commented:
@vedantparanjape-amd The issue should already be fixed by https://github.com/llvm/llvm-project/pull/90489 -- do you still see problems?

nikic reviewed (changes requested):
Already fixed by https://github.com/llvm/llvm-project/pull/90489, removing this from the review queue...
